File: src/ai/models/behavioral-classifier.ts
/**
 * @fileOverview A self-contained Logistic Regression machine learning model in TypeScript.
 * It extracts behavioral features and trains online using the user's local history.
 */

export interface HistoricalTask {
  category: string;
  effort: number;
  scheduledTime: string;
  completed: boolean;
}

export interface PredictionInput {
  category: string;
  effort: number;
  scheduledTime: string;
}

export interface ClassifierPrediction {
  probability: number;
  prediction: 'completed' | 'missed';
  featuresUsed: {
    normalizedEffort: number;
    categoryCompletionRate: number;
    timeOfDayCompletionRate: number;
    previousTaskSuccess: number;
  };
}

export class BehavioralClassifier {
  private weights: number[];
  private bias: number;
  private lr: number = 0.1;
  private epochs: number = 100;

  constructor() {
    // 4 Features:
    // x0: Normalized task effort (1 to 5 maps to 0.0 to 1.0)
    // x1: Historical completion rate for this category
    // x2: Historical completion rate for this hour block (morning, afternoon, evening)
    // x3: Previous task outcome (0 = missed, 1 = completed)
    
    // Initialize weights with negative bias for effort, and positive for consistency features
    this.weights = [-0.3, 0.6, 0.4, 0.3];
    this.bias = -0.1;
  }

  private sigmoid(z: number): number {
    return 1 / (1 + Math.exp(-z));
  }

  /**
   * Determine the time-of-day block: Morning (6-12), Afternoon (12-18), Evening (18-24), Night (0-6)
   */
  private getTimeBlock(timeStr: string): 'morning' | 'afternoon' | 'evening' | 'night' {
    const hour = parseInt(timeStr.split(':')[0]) || 12;
    if (hour >= 6 && hour < 12) return 'morning';
    if (hour >= 12 && hour < 18) return 'afternoon';
    if (hour >= 18 && hour < 24) return 'evening';
    return 'night';
  }

  /**
   * Pre-calculate completion rate aggregates to use as features
   */
  private calculateStats(history: HistoricalTask[]) {
    const categoryTotals: Record<string, number> = {};
    const categoryCompletions: Record<string, number> = {};
    
    const timeTotals: Record<string, number> = { morning: 0, afternoon: 0, evening: 0, night: 0 };
    const timeCompletions: Record<string, number> = { morning: 0, afternoon: 0, evening: 0, night: 0 };

    for (const task of history) {
      // Category Stats
      categoryTotals[task.category] = (categoryTotals[task.category] || 0) + 1;
      if (task.completed) {
        categoryCompletions[task.category] = (categoryCompletions[task.category] || 0) + 1;
      }

      // Time Stats
      const block = this.getTimeBlock(task.scheduledTime);
      timeTotals[block]++;
      if (task.completed) {
        timeCompletions[block]++;
      }
    }

    return {
      categoryTotals,
      categoryCompletions,
      timeTotals,
      timeCompletions,
    };
  }

  /**
   * Extract input features for the classifier
   */
  private extractFeatures(
    task: { category: string; effort: number; scheduledTime: string },
    prevTask: { completed: boolean } | null,
    stats: any
  ): number[] {
    const x0 = (task.effort - 1) / 4; // Normalize effort to [0, 1]

    // Category rate feature
    const catTotal = stats.categoryTotals[task.category] || 0;
    const catComp = stats.categoryCompletions[task.category] || 0;
    const x1 = catTotal > 0 ? catComp / catTotal : 0.5;

    // Time block rate feature
    const block = this.getTimeBlock(task.scheduledTime);
    const timeTotal = stats.timeTotals[block] || 0;
    const timeComp = stats.timeCompletions[block] || 0;
    const x2 = timeTotal > 0 ? timeComp / timeTotal : 0.5;

    // Previous task status
    const x3 = prevTask ? (prevTask.completed ? 1 : 0) : 0.5;

    return [x0, x1, x2, x3];
  }

  /**
   * Train the Logistic Regression classifier on user history using gradient descent
   */
  public train(history: HistoricalTask[]) {
    if (history.length < 3) return; // Require at least 3 historical points to fit parameters

    const stats = this.calculateStats(history);
    const dataset: Array<{ x: number[]; y: number }> = [];

    // Construct training dataset of pairs (features at t, outcome at t)
    for (let i = 1; i < history.length; i++) {
      const task = history[i];
      const prevTask = history[i - 1];
      const x = this.extractFeatures(task, prevTask, stats);
      const y = task.completed ? 1 : 0;
      dataset.push({ x, y });
    }

    // Run Gradient Descent to minimize Binary Cross-Entropy Loss
    for (let epoch = 0; epoch < this.epochs; epoch++) {
      for (const { x, y } of dataset) {
        const z = x.reduce((sum, val, idx) => sum + val * this.weights[idx], 0) + this.bias;
        const yHat = this.sigmoid(z);
        
        // Gradient = yHat - y
        const gradient = yHat - y;

        // Weights & Bias Updates
        for (let j = 0; j < this.weights.length; j++) {
          this.weights[j] -= this.lr * gradient * x[j];
        }
        this.bias -= this.lr * gradient;
      }
    }
  }

  /**
   * Predict completion outcome for a target task
   */
  public predict(target: PredictionInput, history: HistoricalTask[]): ClassifierPrediction {
    const stats = this.calculateStats(history);
    const prevTask = history.length > 0 ? history[0] : null; // Most recent task in sorted history

    const x = this.extractFeatures(target, prevTask, stats);
    const z = x.reduce((sum, val, idx) => sum + val * this.weights[idx], 0) + this.bias;
    const probability = this.sigmoid(z);

    return {
      probability,
      prediction: probability >= 0.5 ? 'completed' : 'missed',
      featuresUsed: {
        normalizedEffort: x[0],
        categoryCompletionRate: x[1],
        timeOfDayCompletionRate: x[2],
        previousTaskSuccess: x[3]
      }
    };
  }

  /**
   * Return the model's trained weights and bias parameters.
   */
  public getModelInfo() {
    return {
      weights: this.weights,
      bias: this.bias,
      featureNames: [
        'Normalized Effort',
        'Category Completion Rate',
        'Time-of-day Completion Rate',
        'Previous Task Success'
      ]
    };
  }
}

File: scripts/behavioral_classifier.py
import numpy as np
import json

class BehavioralClassifier:
    """
    A standalone Logistic Regression machine learning model in Python.
    Replicates the exact feature extraction, gradient descent, and prediction
    logic used in the Next.js TypeScript application.
    """
    def __init__(self):
        # 4 Features:
        # x0: Normalized task effort (1 to 5 maps to 0.0 to 1.0)
        # x1: Historical completion rate for this category
        # x2: Historical completion rate for this hour block (morning, afternoon, evening, night)
        # x3: Previous task outcome (0 = missed, 1 = completed)
        self.weights = np.array([-0.3, 0.6, 0.4, 0.3])
        self.bias = -0.1
        self.lr = 0.1
        self.epochs = 100

    def sigmoid(self, z):
        return 1 / (1 + np.exp(-z))

    def get_time_block(self, time_str):
        try:
            hour = int(time_str.split(':')[0])
        except Exception:
            hour = 12
        
        if 6 <= hour < 12:
            return 'morning'
        elif 12 <= hour < 18:
            return 'afternoon'
        elif 18 <= hour < 24:
            return 'evening'
        else:
            return 'night'

    def calculate_stats(self, history):
        category_totals = {}
        category_completions = {}
        time_totals = {'morning': 0, 'afternoon': 0, 'evening': 0, 'night': 0}
        time_completions = {'morning': 0, 'afternoon': 0, 'evening': 0, 'night': 0}

        for task in history:
            cat = task['category']
            category_totals[cat] = category_totals.get(cat, 0) + 1
            if task['completed']:
                category_completions[cat] = category_completions.get(cat, 0) + 1

            block = self.get_time_block(task['scheduledTime'])
            time_totals[block] += 1
            if task['completed']:
                time_completions[block] += 1

        return {
            'category_totals': category_totals,
            'category_completions': category_completions,
            'time_totals': time_totals,
            'time_completions': time_completions
        }

    def extract_features(self, task, prev_task, stats):
        # Normalize effort from [1, 5] to [0.0, 1.0]
        x0 = (task['effort'] - 1) / 4.0

        # Category Completion Rate
        cat_total = stats['category_totals'].get(task['category'], 0)
        cat_comp = stats['category_completions'].get(task['category'], 0)
        x1 = cat_comp / cat_total if cat_total > 0 else 0.5

        # Time Block Completion Rate
        block = self.get_time_block(task['scheduledTime'])
        time_total = stats['time_totals'].get(block, 0)
        time_comp = stats['time_completions'].get(block, 0)
        x2 = time_comp / time_total if time_total > 0 else 0.5

        # Previous Task Outcome
        x3 = 1.0 if prev_task and prev_task['completed'] else 0.0 if prev_task else 0.5

        return np.array([x0, x1, x2, x3])

    def train(self, history):
        if len(history) < 3:
            print("Model requires at least 3 points to train parameters.")
            return

        stats = self.calculate_stats(history)
        dataset = []

        # Construct training dataset of pairs (features at t, outcome at t)
        for i in range(1, len(history)):
            task = history[i]
            prev_task = history[i - 1]
            x = self.extract_features(task, prev_task, stats)
            y = 1.0 if task['completed'] else 0.0
            dataset.append((x, y))

        # Run Gradient Descent to minimize Binary Cross-Entropy Loss
        for epoch in range(self.epochs):
            for x, y in dataset:
                z = np.dot(x, self.weights) + self.bias
                y_hat = self.sigmoid(z)
                
                # Gradient = prediction error
                gradient = y_hat - y

                # Weights & Bias Updates
                self.weights -= self.lr * gradient * x
                self.bias -= self.lr * gradient

    def predict(self, target, history):
        stats = self.calculate_stats(history)
        prev_task = history[0] if len(history) > 0 else None

        x = self.extract_features(target, prev_task, stats)
        z = np.dot(x, self.weights) + self.bias
        probability = self.sigmoid(z)

        return {
            'probability': float(probability),
            'prediction': 'completed' if probability >= 0.5 else 'missed',
            'featuresUsed': {
                'normalizedEffort': float(x[0]),
                'categoryCompletionRate': float(x[1]),
                'timeOfDayCompletionRate': float(x[2]),
                'previousTaskSuccess': float(x[3])
            }
        }

if __name__ == "__main__":
    # Test Data: historical logs
    mock_history = [
        {"category": "work", "effort": 3, "scheduledTime": "09:00", "completed": True},
        {"category": "work", "effort": 4, "scheduledTime": "10:30", "completed": False},
        {"category": "health", "effort": 2, "scheduledTime": "07:00", "completed": True},
        {"category": "learning", "effort": 5, "scheduledTime": "20:00", "completed": False},
        {"category": "work", "effort": 3, "scheduledTime": "14:00", "completed": True},
        {"category": "personal", "effort": 1, "scheduledTime": "18:00", "completed": True},
        {"category": "work", "effort": 3, "scheduledTime": "09:00", "completed": True},
    ]

    print("=== Behavioral Classifier ML Script (Python) ===")
    classifier = BehavioralClassifier()
    print("\nInitial parameters:")
    print(" - Weights:", classifier.weights)
    print(" - Bias:   ", classifier.bias)

    # Train model
    print("\nTraining on", len(mock_history), "historical logs...")
    classifier.train(mock_history)
    
    print("\nTrained parameters:")
    print(" - Weights:", classifier.weights)
    print(" - Bias:   ", classifier.bias)

    # Run a prediction
    target_task = {"category": "work", "effort": 3, "scheduledTime": "09:00"}
    prediction = classifier.predict(target_task, mock_history)
    print("\nPrediction for task:", target_task)
    print(json.dumps(prediction, indent=2))
